### PR TITLE
Add support for publishing events among plugins

### PR DESCRIPTION
**Pull request by JohnNiang**
#### What type of PR is this?

/kind feature
/area core
/area plugin

#### What this PR does / why we need it:

This PR enhance usage of SharedEvent annotation to add support for publishing events among plugins.

#### How to test?

1. Clone repository https://github.com/halo-dev/plugin-starter
2. Change build.gradle as following:

	```gradle
	dependencies {
	    implementation platform('run.halo.tools.platform:plugin:2.17.0-SNAPSHOT')
	```

3. Change StarterPlugin as following:

	```java
	@Component
	public class StarterPlugin extends BasePlugin {
	
	    private final ApplicationContext appContext;
	
	    public StarterPlugin(PluginContext pluginContext, ApplicationContext appContext) {
	        super(pluginContext);
	        this.appContext = appContext;
	    }
	
	    @Override
	    public void start() {
	        appContext.publishEvent(new PostDeletedEvent(this, "fake-plugin"));
	    }
	
	    @Override
	    public void stop() {
	    }

	    @EventListener(PostDeletedEvent.class)
	    public void onApplicationEvent(PostDeletedEvent event) {
	        System.out.println("Post deleted event received in plugin: " + event.getName());
	    }
	}
	```
4. Add a listener to Halo core
```java
	    @EventListener(PostDeletedEvent.class)
	    public void onApplicationEvent(PostDeletedEvent event) {
	        System.out.println("Post deleted event received in core: " + event.getName());
	    }
```
5. Build plugin and install plugin
6. Enable the plugin and see the result

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by ruibaby** [comment]
@JohnNiang 是否需要补充一下插件开发的文档？

**Comment by ruibaby** [comment]
提个 PR 到 halo-dev/docs 吧。

https://github.com/halo-dev/halo/pull/6081/commits/8c40bf467867fde76d9e0966307089d4cbb60f9d#diff-b8447709f93eeedfadb44f95a9c5fd9f508c8a48c446b9df0f5022722df2992d

File: docs/plugin/shared-event.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    super(source);
+  }
+
+}
+```
+
+> 需要注意的是：
+> 1. 共享事件必须继承 `ApplicationEvent`。
+> 2. 共享事件必须被 `@SharedEvent` 注解标记。
+> 3. 如果想要被其他插件订阅，则需要将改事件类发布到 Maven 仓库中，供其他插件引用。
```
Comment (guqing) [inline docs/plugin/shared-event.md:83]:
```suggestion
> 3. 如果想要被其他插件订阅，则需要将该事件类发布到 Maven 仓库中，供其他插件引用。
```

**Review by ruibaby — COMMENTED** [review]
/lgtm

**Review by guqing — APPROVED** [review]
/approve

**Pull request merged**